### PR TITLE
Python-only build script

### PR DESCRIPTION
**Pull request by drodriguez**
Applies to [SR-237](https://bugs.swift.org/browse/SR-237).

This is a partial reimplementation of `build-script-impl` into Python. This was necessary if we wanted to build on Windows with the `build-script`. At the same time it includes fixes for compatibility with Python 3 (keeping code backwards compatible with Python 2).

Currently it can build and test CMark, LLVM, Swift and LLDB (the test fail in Windows, but that's not the point).

I tried to follow the design that I can intuited in the current Python code, but I have to diverge in some points (because I wanted to keep the usage of `build-script-impl` untouched during the migration). A new parameter `--native-impl` allows to use the new code while still allowing the usage of the old code.

I hope the ideas I introduce makes sense and we can move forward with this. It would be great to have the Windows builds using the same process than the rest of the builds, and I think it would be a requirement for being integrated with the community CI. This rewrites will also help to future modifications of the build script, since information and process will not have to be duplicated in two places.

Any feedback is appreciated.

### Changes

- Move `HostSpecificConfiguration` into its own file to be shared between the main script and the products that need to make use of it to recover CMake targets (mostly Swift, at the moment). There's also changes to avoid depending on the invocation object directly and be less tied to it.
- As a transitional measure, much of the code of `BuildScriptInvocation` has been moved into `BuildScriptBase`, and it is shared with its subtypes `BuildScriptInvocation` and `BuildScriptNative`.  `Base` will invoke different (abstract) methods, that the subtypes implement to either invoke `build-script-impl` with the right arguments, or start executing some Python code for each product.
- There is the new concept of "product builders", which implement the logic to build the products. Each product provides a class method to recover its product builder (or in the case of LLDB, one of the two possible product builders). These product builders receive the arguments, the toolchain, the workspace, the host,... and take care of massaging all those values into proper CMake or shell invocations.
- A template for a builder using CMake is provided: `CMakeProductBuilder` (currently in `products/product.py`, but should be moved out). It implements the common logic for building with CMake, and provides many templated methods/properties for the products to customize. CMark, LLVM, Swift and LLDB use this builder as a base.
- The `CMake` type learns a couple new tricks and help invoking CMake for generating the build files, and for both building targets and installing targets.
- Some parameters have been copied from `build-script-impl` to `build-script` in order to get access to their values during the build process. All of them should be forwarded to `build-script-impl`. When the arguments where used in one of the build presets, I tried to keep compatibility with how the `build-script-impl` argument seemed to work.
- I repurposed the parameters used for passing ICU paths for Android for Windows. This might not be very clever if we ever want to cross-compile Android from Windows, but for the time being, it works.
- In many places I needed to modify how Python scripts were invoked for Windows. In some cases, a `.cmd` wrapper is provided (`build-script.cmd`, `coverage-touch-tests.cmd`), but in other cases, Python is invoked internally (like `llvm-lit.py`).
- Many changes for Windows compatibility.
- Many changes for Python 3 compatibility.

### TODO
- [ ] Transform the rest of the products to use CMake builder or the appropiate builder.
- [ ] `install` actions do not work. `merge`, `extractsymbols` and `package` do not work either.
- [ ] Find out about many of the arguments that do not seem to be used in `build-script-impl`. Is it worth it to bring them over, or can the code be removed?
- [ ] Disable the invocation of `build-script-impl` to check the arguments if using the native implementation. It will not work on Windows (for example). I have a workaround locally, but it should be encoded into the script (no need to check the arguments against `impl` if we are not going to use it).
- [ ] Modify Ninja, indexstore-db and sourcekit-lsp to adapt to the new idea of the product builders.
- [ ] Check this script on Linux
- [ ] Check this script on Darwin
- [ ] Check all the build in Python 2.

/cc @Rostepher @shahmishal 
/cc @benlangmuir: I saw that you modified the code recently to build indexstore-db and sourcekit-lsp. I think the new design I propose will work with those changes. If you have any feedback, please don't hesitate to comment.

### CODE REVIEW

**Comment by openloop** [comment]
Are TODOs in FIFO order? Make is pretty important. Nice writeup.

**Comment by benlangmuir** [comment]
> I saw that you modified the code recently to build indexstore-db and sourcekit-lsp. I think the new design I propose will work with those changes. If you have any feedback, please don't hesitate to comment.

The only "interesting" part of building these is that you will need support for *installing* the earlier parts of the toolchain.

**Comment by drodriguez** [comment]
@openloop: the order was as they were coming to my mind. No other intentional order.

@benlangmuir: All right. I wasn't sure why they were done after the rest of the projects had finished the build-test-install steps. That's something I will have to had into account. Thanks for pointing it out. Do you remember why they have to be installed? Dispatch and Foundation use Swift in the build directory without first installing it. Can those other do the same?

**Comment by Rostepher** [comment]
As a quick note, I think non-`build-script` changes should be split out of this PR. I see a lot of changes to various `.swift.gyb` files and scripts in `utils`. I'd recommend splitting those updates out into separate PRs to avoid making one massive PR that becomes un-reviewable and worse yet un-testable.

**Comment by drodriguez** [comment]
@Rostepher: Totally agree. There's a mix here between compatibility with Python 3, compatibility with Windows and the new build script. Most of the changes for Python 3 should be sent independently. I'm working in all at the same time because it makes sense to have everything in the same branch (for me). But I will have to split them in smaller PRs. And I will do it when I'm sure they are mostly stable and unchanging.

**Comment by benlangmuir** [comment]
> Do you remember why they have to be installed? Dispatch and Foundation use Swift in the build directory without first installing it. Can those other do the same?

Passing through all of the paths doesn't scale, particularly for tools that are downstream of swiftpm or Foundation.  There's more information at https://forums.swift.org/t/rfc-building-swift-packages-in-build-script/18920

**Comment by drodriguez** [comment]
That write up is great. Thanks for sharing. I wonder why both `build-script-impl` and `build-script` try to do all build phases, then all test phases, and then all install phases. It would be simpler to do build/test/install for each product. That way you will have a simple dependency tree. Did you figure out if there is a reason for it? (I can imagine that building is faster than testing to find early problems, but doesn't feel like a strong reason).

**Comment by benlangmuir** [comment]
I don't know if there is a principled reason for the current state of affairs.  I suggest not trying to change the order of operations at the same time as rewriting the infrastructure.

File: utils/coverage/coverage-touch-tests [hunk 1]
```diff
@@ -86,7 +86,7 @@ def main():
     console.setLevel(level=args.log.upper())
     logging.debug(args)
     environment = os.environ.copy()
-    environment['GIT_EXTERNAL_DIFF'] = script_path
+    environment['GIT_EXTERNAL_DIFF'] = script_path.replace('\\', '/')
```
Comment (compnerd) [inline utils/coverage/coverage-touch-tests]: I could've sworn that that there was a path utility to convert the slashes in python.

Comment (Rostepher) [reply utils/coverage/coverage-touch-tests]: It looks like [`os.path.normpath`](
https://docs.python.org/2/library/os.path.html#os.path.normpath) does what you're looking for.

Comment (drodriguez) [reply utils/coverage/coverage-touch-tests]: It does the opposite if I read the documentation right: “On Windows, it converts forward slashes to backward slashes.”. In this case, for some reason, `GIT_EXTERNAL_DIFF` was removing the backward slashes from the path, but worked when I provided the forward slashes. I can add a comment explaining why it is needed.

**Comment by drodriguez** [comment]
Abandoning this efforts. It might still be useful for better Windows builds for most of the community, but I will not have time to keep improving this and it doesn't seem to be a lot of interest in improving the build system in this direction.

**Pull request closed without merging**